Fix and update dotnet-core and add powershell

mwrock opened the pull request:
This PR was originally to build coreclr from source. I got a long way down that road and learned alot but ended up discovering the value is not worth the cost. The cost being many hours to build the native clr binaries and eventually in the later artifacts, dotnet installs dotnet to build the managed dotnet stack. So it would be a long trip down a rabbit hole to unravel all of that.

Properly relocating the built bits, a binary plan should work fine and runs the accompanying powershell source build on moby linux succesfully.

One change this takes from the previous donet-core plan is to more aggressively patchelf the binaries instead of using the wrapper to set a LD_LIBRARY_PATH. That strategy works when using the dotnet cli but the built .net artifacts will break if calling them directly, which is common.

mwrock commented on dotnet-core/plan.sh:
```diff
@@ -1,58 +1,113 @@
+pkg_shasum=b49ba545fe632dfd5426669ca3300009a5ffd1ccf3c1cf82303dcf44044db33d
+pkg_deps=()
+pkg_build_deps=(
+  core/which
```
Some of these deps are unnecessary and will be dropped

mwrock commented:
Disclaimer: before starting this PR I did not know what `ld` did so thats where my linux c++ tool chain knowledg is.

Currently this is configuring and the compilation fails ~30% of the way in because it complains that "this platform" does not support hi-res time. I looked into the cmake script where it makes that determination. It basically wants to see if `clock_gettime` works. To od so it creates, compiles and runs:
```
#include <stdlib.h>
#include <time.h>
#include <sys/time.h>

int main()
{
	  int ret;
         struct timespec ts;
         ret = clock_gettime(CLOCK_REALTIME, &ts);
	 exit(ret);
}
```
It wants a `0` exit.

I'm running in docker on moby linux. I have manually compiled this and it succeeds ONLY IF I `patchelf` the built binary and change the interpretter. I have also looked at the headers of the libraries that the build creates in its first 30% and their interpretter is pointing to `/lib64/ld-linux-x86-64.so.2` and this is why I think the cmake test is bombing. So I'm looking into how to get the code built with the right interpretter.

mwrock commented:
I "cheated" and got past this with:
```
ln -sf $(pkg_path_for glibc)/lib/ld-linux-x86-64.so.2 /lib64/ld-linux-x86-64.so.2
```
This allowed the cmake CXX source tests to validate the time functions and now the build makes it to ~86%.

It now dies with a bunch of `undefined reference to `_Ux86_64_get_reg'` and other `_Ux86_*` functions. Pretty sure this is a linking issue with libunwind.

mwrock commented:
ok still dirty but the native core clr build is running to completion. Remaining here:

- [ ] Native core Lib
- [ ] Nuget package download
- [ ] Test binaries actually run
- [ ] Sanitize and cleanup code

mwrock commented on dotnet-core/plan.sh:
```diff
@@ -25,34 +24,35 @@ pkg_deps=(
-  patchelf --interpreter "$(pkg_path_for glibc)/lib/ld-linux-x86-64.so.2" \
-    --set-rpath "$LD_RUN_PATH" \
-    ./dotnet
+  hab pkg install core/icu/52.1
```
Had to do this to avoid a conflict with `core/icu`'s gcc-libs. I think rebuilding core/icu will fix that and then I can just include it as a normal dep.

reset commented on dotnet-core/plan.sh:
```diff
@@ -25,34 +25,33 @@ pkg_deps=(
   core/patchelf
 )
-pkg_bin_dirs=(bin)
+pkg_bin_dirs=(.)
```
I see why this has been set to `(.)`, but we should move the binaries that get created into a `bin` directory and set this back to `(bin)`

mwrock replied:
sounds good, will just throw everything in bin

reset commented on dotnet-core/plan.sh:
```diff
@@ -25,34 +25,33 @@ pkg_deps=(
 
 do_install() {
-  # Copy the files into the package
   cp -a . "$pkg_prefix"
```
This is the spot where you'd install the binaries to the `bin` directory

reset commented on powershell/plan.sh:
```diff
@@ -0,0 +1,116 @@
+  core/make
+  core/git
+)
+pkg_bin_dirs=(.)
```
Same comments here as above

reset commented on powershell/plan.sh:
```diff
@@ -0,0 +1,116 @@
+
+  # The build expects to find this file to determine
+  # platform so we will fake ubuntu 14.04
+  cat > "/etc/os-release" <<EOF
```
What does powershell exactly do with this? We might want to specify something Habitat related instead of faking Ubuntu if that's possible

smith replied:
This is "probably fine" in a studio but run anywhere where this file actually does exist it will replace the contents. Probably not cool if it's not an Ubuntu 14.04 box already; could cause all kinds of chaos. We should probably find a better way to do this. Overwriting files in /etc is something we really should avoid in a plan.

mwrock replied:
I think there is a better way around this. I can inject `linux.-x64` into the resgen `project.json`'s supported `runtimes` array and probably eliminate the need for this file alltogether.

smith commented on powershell/plan.sh:
```diff
@@ -0,0 +1,116 @@
+  # This is why we do not download source
+  git clone -b v$pkg_version --recursive "$pkg_source"
+  popd
+  return 0
```
You don't need to `return 0` here. We put those in when doing empty functions because bash does not allow empty function bodies. The function will return the return code of the last command in it.

smith commented on powershell/plan.sh:
```diff
@@ -0,0 +1,116 @@
+
+do_prepare() {
+  cp -r $HAB_CACHE_SRC_PATH/PowerShell/* $HAB_CACHE_SRC_PATH/$pkg_dirname
+  rm -rf $HAB_CACHE_SRC_PATH/PowerShell
```
Do we need to delete this or will `do_clean` do that? Should we be using a version-specific dir instead of just PowerShell?

mwrock replied:
yeah this surprised me. I expected `clean` to zap this but it didn't and every rebuild complained about it.

mwrock commented on dotnet-core/plan.sh:
```diff
@@ -25,34 +25,33 @@ pkg_deps=(
-EOF
-  chmod +x "$pkg_prefix/bin/dotnet"
+do_check() {
+  ./dotnet --version
```
Yeah we can do a:
```
dotnet new
dotnet run
```
which creates a hello world console app and runs it. That should be sufficient.

mwrock commented:
@smith I juts ran a `get-command` which did work but based on the output there are a bunch of missing commands like the ones you ran into. Looking more closely it looks like powershell is not loading any modules at all like even the ones you would expect to be loaded by default such as `Microsoft.Powershell.Management`. Need to investigate that.

mwrock commented:
This morning's commit addresses all comments except why powershell is not importing the "built in" modules. I have confirmed that this only happens in the dockerized studio. I see all commands in a non studio environment as well as a plain chrooted linux studio. It may end up that I'll need to add a wrapper that manually imports these modules. I'll be investigating later why powershell is not auto loading these modules.

mwrock commented:
All comments should now be addressed. The biggest issue was getting the powershell drive path resolution to work in a dockerized studio. Basically .Net searches for mount points and uses that metadata to create a `/` filesystem drive. Because the docker studio environment has no `/` mount point, powershell creates a `/` drive that points to `/dev` and causes many internal powershell path resolution issues resulting in a failure to load any module which means many of powershell's core commands are missing.

I have added a patch to powershell to fix that and also submitted it upstream: https://github.com/PowerShell/PowerShell/pull/2679

mwrock commented:
packages:
```
core/dotnet-core/1.0.0-preview3-003930/20161115063144
core/powershell/6.0.0-alpha.12/20161115064801
```
have been uploaded to the depot.
